libdnf/dnf-sack.cpp

```diff
@@ -611,11 +604,28 @@ write_ext(DnfSack *sack, HyRepo hrepo, _hy_repo_repodata which_repodata,
         FILE *fp = fdopen(tmp_fd, "w+");
 
         g_debug("%s: storing %s to: %s", __func__, repo->name, tmp_fn_templ);
-        if (which_repodata != _HY_REPODATA_UPDATEINFO)
-            ret |= repodata_write(data, fp);
-        else
-            ret |= write_ext_updateinfo(hrepo, data, fp);
-        ret |= checksum_write(repoImpl->checksum, fp);
+
+        DnfSolvfileUserdata dsu;
+        ret |= !dnf_solvfile_userdata_fill(&dsu, repoImpl->checksum);
+        Repowriter *writer = repowriter_create(repo);
+        repowriter_set_userdata(writer, &dsu, DNF_SOLVFILE_USERDATA_BYTES);
+        if (which_repodata != _HY_REPODATA_UPDATEINFO) {
+            repowriter_set_repodatarange(writer, data->repodataid, data->repodataid + 1);
+            repowriter_set_flags(writer, REPOWRITER_NO_STORAGE_SOLVABLE);
+            ret |= repowriter_write(writer, fp);
+        } else {
+            // write only updateinfo repodata
+            int oldstart = repo->start;
+            repo->start = repoImpl->main_end;
+            repo->nsolvables -= repoImpl->main_nsolvables;
+            repowriter_set_flags(writer, REPOWRITER_LEGACY);
+            repowriter_set_keyfilter(writer, write_ext_updateinfo_filter, data);
+            repowriter_set_keyqueue(writer, 0);
+            ret |= repowriter_write(writer, fp);
+            repo->start = oldstart;
+            repo->nsolvables += repoImpl->main_nsolvables;
+        }
+        repowriter_free(writer);
```

Comment:
I assume you have to do this sort of low-level approach due to having to use the repowriter? It's not apparent this does the same thing the old code was doing...

Reply:
Yes, I had to inline a couple of functions (among them even `repo_write_filtered()` from libsolv, but it is deprecated anyway) to get to the `repowriter`.